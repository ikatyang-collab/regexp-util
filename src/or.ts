import { Base } from './base.js'
import { Charset, CharsetInput } from './charset.js'
import { wrap } from './util.js'

export type OrInput = string | Or | CharsetInput

export class Or extends Base {
  public charset: Charset
  public strings: string[]

  constructor(...inputs: OrInput[]) {
    super()

    const strings: string[] = []
    const charset_inputs: CharsetInput[] = []

    for (const input of inputs) {
      if (input instanceof Or) {
        strings.push(...input.strings)
        charset_inputs.push(input.charset)
      } else if (
        typeof input !== 'string' ||
        (input.length === 1 && !is_special_char(input))
      ) {
        charset_inputs.push(input)
      } else {
        strings.push(input)
      }
    }

    this.strings = Object.keys(dictionaryify(strings))
    this.charset = new Charset(...charset_inputs)
  }

  public union(...inputs: OrInput[]) {
    return new Or(this.charset, ...this.strings, ...inputs)
  }

  public subtract(...inputs: OrInput[]) {
    const string_dictionay = dictionaryify(this.strings)
    const charset = new Charset(...this.charset.data)

    const charset_inputs: CharsetInput[] = []
    for (const input of inputs) {
      if (input instanceof Or) {
        for (const str of input.strings) {
          delete string_dictionay[str]
        }
        charset_inputs.push(input.charset)
      } else if (
        typeof input !== 'string' ||
        (input.length === 1 && !is_special_char(input))
      ) {
        charset_inputs.push(input)
      } else {
        delete string_dictionay[input]
      }
    }

    return new Or(
      charset.subtract(...charset_inputs),
      ...Object.keys(string_dictionay),
    )
  }

  protected _is_empty() {
    return this.charset.isEmpty() && this.strings.length === 0
  }

  protected _to_string() {
    const parts: string[] = []

    if (this.charset.data.length !== 0) {
      parts.push(this.charset.toString())
    }

    if (this.strings.length !== 0) {
      parts.push(...this.strings)
    }

    return wrap(parts.join('|'))
  }
}

export const or = (...inputs: OrInput[]) => new Or(...inputs)

function is_special_char(char: string) {
  switch (char) {
    case '^':
    case '$':
      return true
    default:
      return false
  }
}

function dictionaryify(array: string[]) {
  const dictionary: Record<string, true> = {}
  for (const value of array) {
    dictionary[value] = true
  }
  return dictionary
}
