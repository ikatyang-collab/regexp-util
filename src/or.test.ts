import { test, expect } from 'vitest'
import { or } from './or.js'

test('create: accept string', () => {
  expect(or('hello', 'world').toString()).toMatchSnapshot()
})

test('create: accept charset inputs', () => {
  expect(or('a', [1, 3]).toString()).toMatchSnapshot()
})

test('create: accept or', () => {
  expect(or('a', [1, 3], or('c', 'd')).toString()).toMatchSnapshot()
})

test('create: special chars are considered string', () => {
  expect(or('a', '^', '$').toString()).toMatchSnapshot()
})

test('union', () => {
  expect(
    or('hello', 'world', [5, 9]).union([3, 7]).toString(),
  ).toMatchSnapshot()
})

test('subtract: default', () => {
  expect(
    or('hello', 'world', [1, 7], 'foo', 'bar')
      .subtract('world', [3, 5], or('bar', 'baz'))
      .toString(),
  ).toMatchSnapshot()
})

test('subtract: special chars are considered string', () => {
  expect(or('a', '^', '$').subtract('^').toString()).toMatchSnapshot()
})
