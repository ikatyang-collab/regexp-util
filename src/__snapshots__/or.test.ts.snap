// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create: accept charset inputs 1`] = `"(?:[\\\\u0001-\\\\u0003\\\\u0061])"`;

exports[`create: accept or 1`] = `"(?:[\\\\u0001-\\\\u0003\\\\u0061\\\\u0063-\\\\u0064])"`;

exports[`create: accept string 1`] = `"(?:hello|world)"`;

exports[`create: special chars are considered string 1`] = `"(?:[\\\\u0061]|^|$)"`;

exports[`subtract: default 1`] = `"(?:[\\\\u0001-\\\\u0002\\\\u0006-\\\\u0007]|hello|foo)"`;

exports[`subtract: special chars are considered string 1`] = `"(?:[\\\\u0061]|$)"`;

exports[`union 1`] = `"(?:[\\\\u0003-\\\\u0009]|hello|world)"`;
